### PR TITLE
Attention unet

### PR DESCRIPTION
**Pull request by andreanne-lemay**
Add attention unet on 3D Unet and Visualisation of last attention layer

### CODE REVIEW

File: config/config_tumorSeg.json [hunk 1]
```diff
@@ -10,31 +10,32 @@
     "contrast_test": ["T1w","T2w"],
     "center_test": [],
     "retrain_model": null,
-    "batch_size": 1,
+    "batch_size": 2,
```
Comment (charleygros) [inline config/config_tumorSeg.json]: youhou

File: config/config_tumorSeg.json [hunk 2]
```diff
@@ -49,19 +50,12 @@
         "sigma_range":  [3.5, 4.5],
         "p": 0.1
       },
-      "RandomAffine3D": {
-        "degrees": 4.6,
-        "translate": [0.03, 0.03],
-        "scale": [0.98, 1.02]
-      },
```
Comment (charleygros) [inline config/config_tumorSeg.json]: Curious to know why at some point :-)

File: ivadomed/main.py [hunk 3]
```diff
@@ -319,7 +321,8 @@ def cmd_train(context):
                   mt_metrics.specificity_score,
                   mt_metrics.intersection_over_union,
                   mt_metrics.accuracy_score]
-
+    att_maps = list()
+    int_imgs = list()
```
Comment (charleygros) [inline ivadomed/main.py]: Could you please add a short comment here? something like "int_imgs stores XX"

File: ivadomed/models.py [hunk 4]
```diff
@@ -331,11 +327,12 @@ class UNet3D(nn.Module):
     instead of BatchNorm due to small batch size in 3D and the addition of segmentation layers in the decoder.
```
Comment (charleygros) [inline ivadomed/models.py]: We could also add the Attention Gate as new features here!

File: ivadomed/models.py [hunk 5]
```diff
@@ -561,7 +592,7 @@ def forward(self, x):
         out = self.norm_lrelu_upscale_conv_norm_lrelu_l3(out)
 
         # Level 4 localization pathway
-        out = torch.cat([out, context_1], dim=1)
+        out = torch.cat([context_1, out], dim=1)
```
Comment (charleygros) [inline ivadomed/models.py]: Oh! Does it matter? Curious about this :-)

File: ivadomed/models.py [hunk 6]
```diff
@@ -577,3 +608,146 @@ def forward(self, x):
...
+        '''
+        :param x: (b, c, t, h, w)
+        :param g: (b, g_d)
+        :return:
+        '''
+
+        output = self.operation_function(x, g)
+        return output
+
+    def _concatenation(self, x, g):
```
Comment (charleygros) [inline ivadomed/models.py]: I wonder if the issue we were discussing this morning could come from somewhere here?!

File: ivadomed/models.py [hunk 7]
```diff
@@ -577,3 +608,146 @@ def forward(self, x):
...
+    if classname.find('Conv') != -1:
+        init.kaiming_normal_(m.weight.data, a=0, mode='fan_in')
+    elif classname.find('Linear') != -1:
+        init.kaiming_normal_(m.weight.data, a=0, mode='fan_in')
+    elif classname.find('BatchNorm') != -1:
+        init.normal_(m.weight.data, 1.0, 0.02)
+        init.constant_(m.bias.data, 0.0)
+
+
+class UnetGridGatingSignal3(nn.Module):
```
Comment (charleygros) [inline ivadomed/models.py]: Not sure to follow you here. Could you please add some docstrings to detail the function of these modules? Many thanks :-)

File: ivadomed/utils.py [hunk 8]
```diff
@@ -457,3 +447,111 @@ def cuda(input_var):
...
+        h_inp.remove()
+        h_out.remove()
+
+        # Rescale the feature-map if it's required
+        if self.upscale: self.rescale_output_array(x.size())
+
+        return self.inputs, self.outputs
+
+
+def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
```
Comment (charleygros) [inline ivadomed/utils.py]: Cool!

File: ivadomed/utils.py [hunk 9]
```diff
@@ -457,3 +447,111 @@ def cuda(input_var):
         return [t.cuda() for t in input_var]
     else:
         return input_var.cuda()
+
+
+class HookBasedFeatureExtractor(nn.Module):
```
Comment (charleygros) [inline ivadomed/utils.py]: Wow! Warning: impressing work here!

File: ivadomed/utils.py [hunk 10]
```diff
@@ -457,3 +447,111 @@ def cuda(input_var):
...
+    # https://gitship.com/neuroscience/nibabel/blob/master/nibabel/orientations.py
+    ref_orientation = nib.orientations.io_orientation(nib_ref.affine)
+    ras_orientation = nib.orientations.io_orientation(nib_ref_canonical.affine)
+    # Return the orientation that transforms from ras to ref_orientation
+    trans_orient = nib.orientations.ornt_transform(ras_orientation, ref_orientation)
+    # apply transformation
+    return nib.orientations.apply_orientation(arr_ras, trans_orient)
+
+
+def save_feature_map(batch, layer_name, context, model, test_input, slice_axis):
```
Comment (charleygros) [inline ivadomed/utils.py]: Really really cool

File: ivadomed/models.py (hunk 5 above)
Comment (andreanne-lemay) [reply ivadomed/models.py]: Nope,  I saw both implementations and both work the same according to my tests :)

File: ivadomed/main.py (hunk 3 above)
Comment (andreanne-lemay) [reply ivadomed/main.py]: Oups this shouldn't be here, it is not used!

**Pull request merged**